File: NetFramework/TreeView/CS/HyperlinksInNodes/HyperLinkPresenter.cs
using C1.Framework;
using C1.Win.TreeView;
using System;
using System.Drawing;
using System.Windows.Forms;

namespace HyperlinksInNodes
{
    public class HyperLinkPresenter : CustomContentPresenter
    {
        private HyperLinkElement _hyperLink;        

        public HyperLinkPresenter()
        {                        
            _hyperLink = new HyperLinkElement();            
            _hyperLink.Style = new Style();
            _hyperLink.Style.Margins = new Thickness(0,4);
            _hyperLink.Click += hyperLink_Click;
            // CustomContentPresenter child has a parent size, so we add an intermediate panel.
            var panel = new RowPanel();
            panel.Children.Add(_hyperLink);

            Child = panel;
        }

        private void hyperLink_Click(object sender, EventArgs e)
        {
            var url = ((Hyperlink)Node.GetValue()).Url;
            System.Diagnostics.Process.Start(url);            
        }

        #region ICellContentPresenter
        public override void SetStyle(TreeNodeCellStyles styles)
        {
            _hyperLink.Style.Font = styles.Font;            
        }

        public override void SetValue(object value)
        {
            _hyperLink.Text = value.ToString();
        }

        public override string ToolTipText
        {
            get
            {
                return string.Empty;
            }
        }

        public override void SetFormat(string format)
        {

        }        
        #endregion
    }

    public class HyperLinkElement : TextElement
    {
        private bool _isVisited;
        private Cursor _oldCursor;
        private Color _defaultColor;

        public HyperLinkElement()
        {
            HitTestable = true;
        }

        public override void OnMouseEnter(EventArgs e)
        {
            base.OnMouseEnter(e);
            // saves cursor and changes to hand
            _oldCursor = HostControl.Cursor;
            HostControl.Cursor = Cursors.Hand;
            // set underline
            Style.Font = new Font(Style.Font, FontStyle.Underline);
        }

        public override void OnMouseLeave(EventArgs e)
        {
            base.OnMouseLeave(e);
            // reset cursor
            HostControl.Cursor = _oldCursor;
            // reset font
            Style.Font = new Font(Style.Font, FontStyle.Regular);
        }

        public bool IsVisited
        {
            get { return _isVisited; }
            set {
                _isVisited = value;
                if (_isVisited)
                {
                    _defaultColor = Style.ForeColor;
                    // set visited color
                    Style.ForeColor = System.Drawing.Color.FromArgb(128, 0, 128);
                }
                else
                {
                    Style.ForeColor = _defaultColor;
                }
            }
        }

        public override void OnClick(EventArgs e)
        {
            base.OnClick(e);
            IsVisited = true;            
            if (Click != null)
                Click(this, e);
        }

        public event EventHandler Click;
    }
}

File: NetFramework/TreeView/VB/HyperlinksInNodes/HyperLinkPresenter.vb
Imports C1.Framework
Imports C1.Win.TreeView

Public Class HyperLinkPresenter
    Inherits CustomContentPresenter

    Private _hyperLink As HyperLinkElement

    Public Sub New()
        _hyperLink = New HyperLinkElement()
        _hyperLink.Style = New Style()
        _hyperLink.Style.Margins = New Thickness(0, 4)
        AddHandler _hyperLink.Click, AddressOf hyperLink_Click
        ' CustomContentPresenter child has a parent size, so we add an intermediate panel.
        Dim panel = New RowPanel()
        panel.Children.Add(_hyperLink)

        Child = panel
    End Sub

    Private Sub hyperLink_Click(sender As Object, e As EventArgs)
        Dim url = DirectCast(Node.GetValue(), Hyperlink).Url
        System.Diagnostics.Process.Start(url)
    End Sub

#Region "ICellContentPresenter"
    Public Overrides Sub SetStyle(styles As TreeNodeCellStyles)
        _hyperLink.Style.Font = styles.Font
    End Sub

    Public Overrides Sub SetValue(value As Object)
        _hyperLink.Text = value?.ToString()
    End Sub

    Public Overrides ReadOnly Property ToolTipText() As String
        Get
            Return String.Empty
        End Get
    End Property

    Public Overrides Sub SetFormat(format As String)

    End Sub
#End Region

End Class

Public Class HyperLinkElement
    Inherits TextElement

    Private _isVisited As Boolean
    Private _oldCursor As Cursor
    Private _defaultColor As Color

    Public Sub New()
        HitTestable = True
    End Sub

    Public Overrides Sub OnMouseEnter(e As EventArgs)
        MyBase.OnMouseEnter(e)
        ' saves cursor and changes to hand
        _oldCursor = HostControl.Cursor
        HostControl.Cursor = Cursors.Hand
        ' set underline
        Style.Font = New Font(Style.Font, FontStyle.Underline)
    End Sub

    Public Overrides Sub OnMouseLeave(e As EventArgs)
        MyBase.OnMouseLeave(e)
        ' reset cursor
        HostControl.Cursor = _oldCursor
        ' reset font
        Style.Font = New Font(Style.Font, FontStyle.Regular)
    End Sub

    Public Property IsVisited() As Boolean
        Get
            Return _isVisited
        End Get
        Set
            _isVisited = Value
            If _isVisited Then
                _defaultColor = Style.ForeColor
                ' set visited color
                Style.ForeColor = System.Drawing.Color.FromArgb(128, 0, 128)
            Else
                Style.ForeColor = _defaultColor
            End If
        End Set
    End Property

    Public Overrides Sub OnClick(e As EventArgs)
        MyBase.OnClick(e)
        IsVisited = True
        RaiseEvent Click(Me, e)
    End Sub

    Public Event Click As EventHandler
End Class
